dask_cuda/explicit_comms/dataframe_shuffle.py

```diff
@@ -1,148 +1,191 @@
+                        _concat(
+                            rank_to_out_parts_list[rank][i], ignore_index=ignore_index
+                        )
+                    ]
+    return rank_to_out_parts_list
+
+
+async def local_shuffle(
+    s,
+    workers: Set[int],
+    in_nparts: Dict[int, int],
+    in_parts: List[Dict[int, DataFrame]],
+    rank_to_out_part_ids: Dict[int, List[int]],
+    ignore_index: bool,
+) -> List[DataFrame]:
+    """Local shuffle operation of the already grouped/partitioned dataframes
+
+    This function is running on each worker participating in the shuffle.
 
-async def _shuffle(s, workers, dfs_nparts, dfs_parts, column):
-    """
     Parameters
     ----------
     s: dict
         Worker session state
     workers: set
         Set of ranks of all the participants
-    dfs_nparts: list of dict
-        List of dict that for each worker rank specifices the
-        number of partitions that worker has. If the worker doesn't
-        have any partitions, it is excluded from the dict.
-        E.g. `dfs_nparts[0][1]` is how many partitions of the "left"
-        dataframe worker 1 has.
-    dfs_parts: list of lists of Dataframes
-        List of inputs, which in this case are two dataframe lists.
-    column : label or list, or array-like
-        The bases of the rearrangement.
+    in_nparts: dict
+        dict that for each worker rank specifices the
+        number of partitions that worker has of the input dataframe.
+        If the worker doesn't have any partitions, it is excluded from the dict.
+    in_parts: list of dict of dataframes
+        List of dataframe groups that need to be shuffled.
+    rank_to_out_part_ids: dict
+        dict that for each worker rank specifices a list of partition IDs that
+        worker should return. If the worker shouldn't return any partitions,
+        it is excluded from the dict.
+    ignore_index: bool
+        Ignore index during shuffle.  If ``True``, performance may improve,
+        but index values will not be preserved.
+
+    Returns
+    -------
+    partitions: list of DataFrames
+        List of dataframe-partitions
     """
+    myrank = s["rank"]
+    eps = s["eps"]
     assert s["rank"] in workers
-    df_parts = dfs_parts[0]
-
-    # Trimming such that all participanting workers get a rank within 0..len(workers)
-    trim_map = {}
-    for i in range(s["nworkers"]):
-        if i in workers:
-            trim_map[i] = len(trim_map)
 
-    rank = trim_map[s["rank"]]
-    eps = {trim_map[i]: s["eps"][trim_map[i]] for i in workers if i != s["rank"]}
+    rank_to_out_parts_list = sort_in_parts(
+        in_parts,
+        rank_to_out_part_ids,
+        ignore_index,
+        concat_dfs_of_same_output_partition=True,
+    )
 
-    df = df_concat(df_parts)
+    # Communicate all the dataframe-partitions all-to-all. The result is
+    # `out_parts_list` that for each worker and for each output partition
+    # contains a list of dataframes received.
```

Comment:
Is all-to-all here referring that all workers will communicate with each other? Just asking this because we've been experiencing issues with those cases when using sockets to manage connections among workers.

Reply:
Agree with this, but I also think it makes sense to start with all-to-all for now, and tackle collective-communication optimization as a seperate exercise.  The most important consideration here is that we are not doing all partitions to all partitions :)

Reply:
No problem, I'm just trying to think about the future impact of this change, given the recent experience we had with that. For IB users, I think this would be no problem when using RDMACM, but people without IB may eventually experience issues at larger scale clusters.

cc @quasiben for awareness.

Reply:
Your comment/concern certainly makes sense - Are you just noting this as a general performance concern, or do you think something about the explicit comms version is particularly dangerous?  The upstream version is also using all-to-all communication between workers.  Is there something different here from a comm-layer perspective?

Reply:
@pentschev, I agree this is something we should be aware of when scaling. We might consider introducing a more advance all-to-all approach in the future.

Reply:
I'm not talking about performance, the problem is a bit harder. On Selene we saw several issues due to the number of file descriptors that need to be opened when we have too many connections with UCX. Also, unlike Distributed, we have one process per GPU (instead of one process per node), which means more overall workers to do the all-to-all communication. At one point you'll end up being limited even by number of ports that can be created on a single machine.

But as I said in my previous comment, this is a problem with considerably larger-scale clusters that we need to be aware of. In principle, UCX should resolve that by using RDMACM (if InfiniBand is available), but otherwise we may _eventually_ hit the limitations I wrote just above. For now I'd say we're better off just leaving it as is now, while simply keeping that in mind.